lib/src/strings.dart

```diff
@@ -94,38 +95,100 @@ String _tryYamlEncodeSingleQuoted(String string) {
+    final first = indent + trimmedSplit[0];
+
+    if (trimmedSplit.length > 1) {
+      /// Try folding to match specification:
+      /// * https://yaml.org/spec/1.2.2/#65-line-folding
+      trimmedString =
+          trimmedSplit.foldIndexed(first, (index, previous, current) {
+        if (index == 0) return previous;
+
+        /// If initially empty, this line holds only `\n`. It's okay. Also, if
+        /// this line is just an empty space.
+        ///
+        /// See https://yaml.org/spec/1.2.2/#64-empty-lines
+        if (current.trim().isEmpty) return previous + lineEnding + current;
+
+        var updated = current;
+
+        /// For consecutive non-empty lines, we add a `\n` with a space.
+        ///
+        /// We need to ensure the ACTUAL previous string was non-empty
+        if (trimmedSplit[index - 1].trim().isNotEmpty) {
+          /// We remove indent and check if the string starts with a space.
+          ///
+          /// Apply `\n` for `foo\nbar` but not `foo\n bar`.
+          if (!current.replaceFirst(indent, '').startsWith(' ')) {
+            updated = lineEnding + updated;
+          }
+        }
 
-  if (removedPortion.contains('\n')) {
-    result = '>+\n${' ' * indentation}';
-  } else {
-    result = '>-\n${' ' * indentation}';
+        /// Apply `\n` by default if no operation occurred
+        return previous + lineEnding + updated;
+      });
+    } else {
+      trimmedString = first;
+    }
   }
 
-  /// Duplicating the newline for folded strings preserves it in YAML.
-  /// Assumes the user did not try to account for windows documents by using
-  /// `\r\n` already
-  return result +
-      trimmedString.replaceAll('\n', lineEnding * 2 + ' ' * indentation) +
-      removedPortion;
+  return literalPrefix + trimmedString + stripped;
 }
 
 /// Generates a YAML-safe literal string.
 ///
 /// It is important that we ensure that [string] is free of unprintable
 /// characters by calling [_hasUnprintableCharacters] before invoking this
 /// function.
-String _tryYamlEncodeLiteral(
-    String string, int indentation, String lineEnding) {
-  final result = '|-\n$string';
+String _tryYamlEncodeLiteral(String string, int indentSize, String lineEnding) {
+  final indent = ' ' * indentSize;
+
+  final (literalPrefix, trimmedString, stripped) = _prepareBlockString(
+    '|',
+    string,
+    indent,
+    lineEnding,
+  );
+
+  /// Simplest block style.
+  /// * https://yaml.org/spec/1.2.2/#812-literal-style
+  return literalPrefix + indent + trimmedString + stripped;
+}
 
-  /// Assumes the user did not try to account for windows documents by using
-  /// `\r\n` already
-  return result.replaceAll('\n', lineEnding + ' ' * indentation);
+/// Prepares a string to be encoded either via [_tryYamlEncodeLiteral] or
+/// [_tryYamlEncodeFolded] when in [CollectionStyle.BLOCK].
+(
+  String literalPrefix,
+  String trimmedString,
+  String strippedString
+) _prepareBlockString(
```

Comment:
Are we sure this method is worth the dedupliation?

It seems weird that we do `trimmed.replaceAll('\n', lineEnding + indent)` and then inside `_tryYamlEncodeFolded` we split it on `lineEnding`.

It's unclear what we return from this method too.

I think may we should just let this be duplicated into the two method.

Then `_tryYamlEncodeFolded` can split on `\n` instead of splitting on `lineEnding`, which is probably going to result in the same thing, but hopefully going to be less confusing.

Reply:
That was part of the existing code.

> It seems weird that we do `trimmed.replaceAll('\n', lineEnding + indent)` and then inside `_tryYamlEncodeFolded` we split it on `lineEnding`.
> 
> It's unclear what we return from this method too.
> 
> I think may we should just let this be duplicated into the two method.
> 
> Then `_tryYamlEncodeFolded` can split on `\n` instead of splitting on `lineEnding`, which is probably going to result in the same thing, but hopefully going to be less confusing.

I am getting confused. Help me clarify something. If I do,

```dart
final multiLine = '''
My multiline string
is great
''';
```

Will `Dart` return the `line breaks` as `\r\n` when on Windows or just plain `\n`.

Reply:
I actually don't know, I would suspect it always uses `\n`, because anything just seems crazy :rofl: 

But as far as yaml_edit is concerned I don't see how that's relevant.

If `package:yaml` parses a YAML document that uses `\n\r` as line breaks, it won't included that kind of line breaks in folded or literal strings.

The only way to get that into a string in YAML is to use a double quoted string with escape characters: `"\n\r"`.

I'm like 99% sure that parsing YAML on Windows won't cause multi-line strings to be broken by `\n\r`, they'll always be broken by `\n` once parsed. The actual YAML file on disk should contain `\n\r`, that's correct. But when parsed the `YamlScalar.value` should never be a string that contains `\n\r`, unless it was a double quoted string with an escape sequence.

Does that makes sense? Yes it's easy to get lost, and I'll admit there is a fair 12.3% risk I missed something here :rofl: